### PR TITLE
[YAML] fix javascript UDF output in mapping transforms

### PR DESCRIPTION
**Pull request by Polber**
Add JSObject python conversion to YAML Javascript UDF Mapping transforms.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Polber** [comment]
R: @robertwb

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 1]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
     self.__dict__.update(state)
 
 
+def row_to_dict(row):
+  if (str(type(row).__name__).startswith('BeamSchema_') or
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:103]: I was hoping there was a better way. Fixed.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 2]
```diff
@@ -27,8 +27,31 @@
 from typing import Union
 
 import js2py
+from js2py.base import JsObjectWrapper
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:30]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 3]
```diff
@@ -27,8 +27,31 @@
...
+from js2py.base import PyJsUint32Array
+from js2py.base import PyJsUint8Array
+from js2py.base import PyJsUint8ClampedArray
+from js2py.base import PyJsUndefined
+from js2py.base import to_python
+from js2py.constructors.jsdate import PyJsDate
+from js2py.internals.simplex import JsException
 
 import apache_beam as beam
+from apache_beam import Row
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:54]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 4]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
     self.__dict__.update(state)
 
 
+def row_to_dict(row):
+  if (str(type(row).__name__).startswith('BeamSchema_') or
+      isinstance(row, Row)):
+    row = row._asdict()
+  if isinstance(row, dict):
+    for key, value in row.items():
+      row[key] = row_to_dict(value)
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:108]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 5]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
...
+  elif not isinstance(row, str) and isinstance(row, Iterable):
+    row_list = list(row)
+    for idx in range(len(row_list)):
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:111]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 6]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
...
 
+def row_to_dict(row):
+  if (str(type(row).__name__).startswith('BeamSchema_') or
+      isinstance(row, Row)):
+    row = row._asdict()
+  if isinstance(row, dict):
+    for key, value in row.items():
+      row[key] = row_to_dict(value)
+  elif not isinstance(row, str) and isinstance(row, Iterable):
+    row_list = list(row)
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:110]: Fixed.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 7]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
...
+      PyJsInt16Array,
+      PyJsUint16Array,
+      PyJsInt32Array,
+      PyJsUint32Array,
+      PyJsFloat32Array,
+      PyJsFloat64Array)
+
+  def _js_object_to_py_object(obj):
+    if isinstance(obj, (PyJsNumber, PyJsString, PyJsBoolean)):
+      obj = to_python(obj)
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:136]: Done.

File: sdks/python/apache_beam/yaml/yaml_udf_test.py [hunk 8]
```diff
@@ -60,16 +66,37 @@ def test_map_to_fields_filter_inline_js(self):
         language: javascript
         fields:
           label:
-            callable: "function label_map(x) {return x.label + 'x'}"
+            callable: |
+              function label_map(x) {
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_udf_test.py:70]: This is testing callable specifically. I wanted to make tests that only use 1 of the 3 types (expression, callable, file) so that the failures are granular.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 9]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
...
+    elif isinstance(obj, PyJsDate):
+      obj = obj.to_utc_dt()
+    elif isinstance(obj, (PyJsNull, PyJsUndefined)):
+      return None
+    elif isinstance(obj, PyJsError):
+      raise RuntimeError(obj['message'])
+    elif isinstance(obj, PyJsObject):
+      return {
+          key: _js_object_to_py_object(value['value'])
+          for key,
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:148]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 10]
```diff
@@ -75,28 +99,89 @@ def __setstate__(self, state):
...
     udf_code = FileSystems.open(path).read().decode()
     js = js2py.EvalJs()
     js.eval(udf_code)
-    js_callable = _CustomJsObjectWrapper(getattr(js, name))
-    return lambda __row__: js_callable(__row__._asdict())
+    js_file_callable = _CustomJsObjectWrapper(getattr(js, name))
+    fn = lambda __row__: lambda: js_file_callable(row_to_dict(__row__))
+
+  return lambda __row__: dicts_to_rows(
+      _js_object_to_py_object(_catch_js_errors(fn(__row__))))
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:184]: It does propagate. It extracts a `PyJsError` object and the error is thrown in `_js_object_to_py_object()` which is caught by the existing error handling framework. I thought it was a bit cleaner of an error message because if you have a custom error/exception in the JS script, it gets a bit hidden in the stack trace of JSException as opposed to being the top-level error

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 11]
```diff
@@ -100,16 +80,13 @@ def __setstate__(self, state):
...
   if isinstance(row, dict):
-    for key, value in row.items():
-      row[key] = row_to_dict(value)
+    return {key: row_to_dict(value) for key, value in row.items()}
   elif not isinstance(row, str) and isinstance(row, Iterable):
-    row_list = list(row)
-    for idx in range(len(row_list)):
-      row_list[idx] = row_to_dict(row_list[idx])
+    return [row_to_dict(value) for value in list(row)]
   return row
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:90]: Nit. I'd put the last return in an else clause (as it's a "mutually exclusive cases" kind of setup.)

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 12]
```diff
@@ -75,28 +79,85 @@ def __setstate__(self, state):
...
+    elif isinstance(obj, base.JsObjectWrapper):
+      return _js_object_to_py_object(obj._obj)
+
+    return obj
+
+  def _catch_js_errors(func):
+    try:
+      result = func()
+    except simplex.JsException as e:
+      result = getattr(e, 'mes')
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:136]: why not `e.mes`?

File: sdks/python/apache_beam/yaml/yaml_mapping.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:160]: Catching an exception, turning it into a value, and then re-raising that value elsewhere makes things a lot harder to follow though. E.g. you catch simplex.JsException, get the `mes` attribute (which is presumably of type `PyJsError` so you can detect it later?, and then raise during conversion? 

If you want a simpler stack trace, you can do something like

```
try:
  return func()
except simplex.JsException as exn:
  raise RuntimeError(f"Error evaluating javascript expression: {exn.msg}") from exn
```

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 13]
```diff
@@ -75,28 +79,85 @@ def __setstate__(self, state):
...
+    return result
+
   if expression:
     args = ', '.join(original_fields)
     js_func = f'function fn({args}) {{return ({expression})}}'
-    js_callable = _CustomJsObjectWrapper(js2py.eval_js(js_func))
-    return lambda __row__: js_callable(*__row__._asdict().values())
+    js_expr_callable = _CustomJsObjectWrapper(js2py.eval_js(js_func))
+    fn = lambda __row__: lambda: js_expr_callable(
+        *row_to_dict(__row__).values())
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:144]: Does js_expr_callable take keyword arguments so you don't have to rely on the ordering of `row_to_dict(__row__).values()`? (I think we'll be OK, as long as __row__ has exactly the expected schema, but maybe it is still compatible but has extra fields?)

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 14]
```diff
@@ -75,28 +79,85 @@ def __setstate__(self, state):
...
+      result = getattr(e, 'mes')
+    return result
+
   if expression:
     args = ', '.join(original_fields)
     js_func = f'function fn({args}) {{return ({expression})}}'
-    js_callable = _CustomJsObjectWrapper(js2py.eval_js(js_func))
-    return lambda __row__: js_callable(*__row__._asdict().values())
+    js_expr_callable = _CustomJsObjectWrapper(js2py.eval_js(js_func))
+    fn = lambda __row__: lambda: js_expr_callable(
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:143]: No need for the double-underscore __row__ as you're not worried about polluting a namespace here.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 15]
```diff
@@ -75,28 +79,85 @@ def __setstate__(self, state):
...
   else:
     if not path.endswith('.js'):
       raise ValueError(f'File "{path}" is not a valid .js file.')
     udf_code = FileSystems.open(path).read().decode()
     js = js2py.EvalJs()
     js.eval(udf_code)
-    js_callable = _CustomJsObjectWrapper(getattr(js, name))
-    return lambda __row__: js_callable(__row__._asdict())
+    js_file_callable = _CustomJsObjectWrapper(getattr(js, name))
+    fn = lambda __row__: lambda: js_file_callable(row_to_dict(__row__))
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:157]: It makes it a bit hard to follow to have fn be a lambda that takes a row and returns a callable. If you want this pattern, this could likely be better done with function decorators. But there's also redundancy here as well. I think it'd be cleaner to do something like

```
fn = [your previous logic defining a js -> js evaluation]

def wrapper(row):
  try:
    return dicts_to_rows(_js_object_to_py_object(fn(row_to_dict(row))))
  except simplex.JsException as exn:
    raise RuntimeError(f"Error evaluating javascript expression: {exn.msg}") from exn

return wrapper
```

If you want to split up the errors you could also do

```
def wrapper(row):
  row_as_dict = row_to_dict(row)
  try:
    js_result = fn(row_as_dict)
  except simplex.JsException as exn:
    raise RuntimeError(f"Error evaluating javascript expression: {exn.msg}") from exn
  return dicts_to_rows(_js_object_to_py_object(js_result))
```

**Review by robertwb — APPROVED** [review]
Drop at TODO at least about getting the type inference working.

**Review by damccorm — APPROVED** [review]
LGTM once checks pass
